pkg/virt-launcher/virtwrap/network/podinterface.go

```diff
@@ -103,27 +103,34 @@ func (l *PodInterface) PlugPhase1(vmi *v1.VirtualMachineInstance, iface *v1.Inte
 	if !isExist {
 		err := driver.discoverPodNetworkInterface()
 		if err != nil {
-			return err
+			return createCriticalNetworkError(err)
 		}
 
 		if err := driver.preparePodNetworkInterfaces(); err != nil {
-			log.Log.Reason(err).Critical("failed to prepare pod networking")
+			log.Log.Reason(err).Error("failed to prepare pod networking")
```

Comment:
just to be certain, are any of these instances where critical areas occur capable of passing at some point if they are retried? I want to make sure we always retry when it's reasonable to do so by reporting the error and ultimately re-enqueuing the key on the controller.

Reply:
`preparePodNetworkInterfaces` is doing networking changes on the virt-launcher pod. It may create a bridge, remove the ip&mac from the pod interface etc. Since currently we don't have mechanism to track in what stage the error occurred and we don't revert the network changes that were already done, any error in `preparePodNetworkInterfaces` is unrecoverable.

This PR is trying to restore the behaviour prior phase1 was moved to virt-handler. Back then, in case of `preparePodNetworkInterfaces` error, virt-launcher was crashing. AFAIK, if virt-launcher crashes it won't be resumed and after some time the vmi will be moved to Failed phase. So now phase1 is on virt-handler, moving the vmi to Failed phase seems to be pretty similar to what we had before the bug was introduced.

BTW, if `discoverPodNetworkInterface` returns an error, it should be recoverable since no network changes are done in the discover. I changed discover error to be critical by mistake. I will push a new commit fixing it.

Reply:
Sorry for barging in the middle here, but I would like to ask why we log here if we do not handle the error? I would expect to log where the error is handled, otherwise just add over the existing error data.

Reply:
There is a log and even event where the error is handled. But I find it useful to keep the log here as well for some reasons- 
1. The log prints the file and the line it is invoke from so it will be easier to understand the error came for `phase1`.
2. Since we don't have a stack trace. The whole flow is clearer when you have a log in the source of the error and also in the place handling the error.
3. We are failing after 1 attempt so this log will add maximum 1 extra line to the log.

Reply:
> Since currently we don't have mechanism to track in what stage the error occurred and we don't revert the network changes that were already done, any error in preparePodNetworkInterfaces is unrecoverable.

my expectation is that all flows should be idempotent, meaning we can execute this same logic over and over and the steps that have already taken place are simply a no-op. There's no keeping up of the state or location that a failure occurred, it's simply just retried later on. 

I'm not sure how this works otherwise. There's no way to guarantee we won't call ```preparePodNetworkInterfaces``` more than once for a VMI, so this logic must be idempotent already... and if it's already idempotent, then I'm not sure why we're treating critical errors like an exception here

Reply:
the virt-handler controller has a [mechanism](https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L355) to avoid `network.SetupPodNetworkPhase1` being called more than once in case of success.
Currently in case of failure will retry (of course only after the panics will be removed). This PR separates the possible failures to 2, recoverable and unrecoverable.

If in case of unrecoverable error we retry it can lead to another error in the next try in the good case.
And to undesired behaviour in the bad case.

`preparePodNetworkInterfaces' is doing network changes on the virt-launcher pod, assuming the network is as was created by the cni. If the error in the previous execution happened after some network changes were made, the result is unpredictable.

Reply:
> the virt-handler controller has a mechanism to avoid network.SetupPodNetworkPhase1 being called more than once in case of success.

that's an optimization to prevent a low level performance issue. Every time we call SetupPodNetworkPhase1 it results in a new posix thread being spawned due to how we're entering into the VMI's namespace. the optimization isn't a guarantee though with respects to not calling the setup multiple times. For example, virt-handler could restart and lose memory of that cache then retry SetupPodNetworkPhase1. 

Before that cache mechanism existed it was more common to call SetupPodNetworkPhase1 multiple times. I don't believe any issues occurred as a result other than the inefficient use of posix threads

> `preparePodNetworkInterfaces' is doing network changes on the virt-launcher pod, assuming the network is as was created by the cni. If the error in the previous execution happened after some network changes were made, the result is unpredictable.

Do we understand a sequence of events that occur that lead up to unpredictable results here?  My concern is mostly in *my* lack of understanding around what is unrecoverable here and why. For instance if we could say, "if x fails, then it will never have the opportunity to pass again due to y reason" then I'd feel comfortable I understood the depth of the situation and that we're making the correct assumption here about critical errors.

If we don't fully understand what's occurring here, I'm inclined to treat these errors like any other error. If the error is truly unrecoverable, then the VMI will still eventually fail and time out during startup. 

Also note that the Critial error resulting in a panic is new behavior. I don't think those errors were originally meant to cause an assertion.

Reply:
If the virt-handler crashed, `PlugPhase1` itself has a [mechanism](https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-launcher/virtwrap/network/podinterface.go#L98) to avoid it being invoked again.

Example1 for unrecoverable `preparePodNetworkInterfaces` error. 
1. We run a vm with masquerade binding.
2. `SetupPodNetworkPhase1` is invoked.
3.  `preparePodNetworkInterfaces` create k6t-eth0 bridge and k6t-eth0-nic.
4. `preparePodNetworkInterfaces`  tries to insert nat rules to iptables but fails.
5. `SetupPodNetworkPhase1` returns with an error to the virt-handler controller and the vmi is re-enqueued.
6. `SetupPodNetworkPhase1` is invoked again.
7. `preparePodNetworkInterfaces` fails since k6t-eth0-nic already exists.

Here we will fail again and again until the virt-handler controller (watchdog) will have a timeout waiting for the domain and move the vmi to failed phase.
So in this case yes, fast failing is just an optimization.

Example2 for unrecoverable `preparePodNetworkInterfaces` error. 
1. We run a vm with bridge binding.
2. `SetupPodNetworkPhase1` is invoked.
3.  `preparePodNetworkInterfaces` removes the mac address from eth0 and set and new *random mac*.
4. `preparePodNetworkInterfaces`  [tries to create k6t-eth0 bridge](https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-launcher/virtwrap/network/podinterface.go#L339) and fails.
5. `SetupPodNetworkPhase1` returns with error to the virt-handler controller and the vmi is re-enqueued.
6. `SetupPodNetworkPhase1` is invoked again.
7. `discoverPodNetworkInterface` reads the (random) mac from eth0 and stores in `vif`.
8. `preparePodNetworkInterfaces` is invoked and executed successfully.
9. Result - the *vm* gets the *random mac* from step 3 and therefore may get a random IP address.

In this case, fast failing is not an optimization. It prevents a bug.

Maybe we can try another approach and try to do a cleanup in case `preparePodNetworkInterfaces`. And only if the cleanup fails do the fast failing.

Regarding the panic history - PR #3128 added the panic to the Critical log. We can see in this [link](https://github.com/kubevirt/kubevirt/pull/3128/files#diff-d3ed8eb8cca020f37cab207c079f2627L112), that the same PR also removed the panic from `PlugPhase1`. So the `panic` wasn't added to  `PlugPhase1`as a side effect. It was there intentionally before.

Reply:
i see, so we definitely have established critical paths during network setup that can only be executed once. Your examples make that much more clear, thanks.

At some point we might need to address this differently, and restructure the code so i can be executed multiple times. For now I see your approach here as being the most pragmatic solution though. :+1: